File: sdk/ts/src/issuer-registry.ts
import { CosmWasmClient, SigningCosmWasmClient, ExecuteResult } from "@cosmjs/cosmwasm-stargate";
import {
  IsAuthorizedResponse,
  IssuerResponse,
  IssuersResponse,
  AdminResponse,
} from "./types";

export class IssuerRegistryClient {
  constructor(
    private readonly client: CosmWasmClient,
    public readonly contractAddress: string,
  ) {}

  // ── Queries ──────────────────────────────────────────────────────

  async isAuthorized(issuer: string, credentialType: string): Promise<IsAuthorizedResponse> {
    return this.client.queryContractSmart(this.contractAddress, {
      is_authorized: { issuer, credential_type: credentialType },
    });
  }

  async getIssuer(issuer: string): Promise<IssuerResponse> {
    return this.client.queryContractSmart(this.contractAddress, {
      issuer: { issuer },
    });
  }

  async listIssuers(startAfter?: string, limit?: number): Promise<IssuersResponse> {
    return this.client.queryContractSmart(this.contractAddress, {
      list_issuers: { start_after: startAfter, limit },
    });
  }

  async getIssuersByType(
    credentialType: string,
    startAfter?: string,
    limit?: number,
  ): Promise<IssuersResponse> {
    return this.client.queryContractSmart(this.contractAddress, {
      issuers_by_type: { credential_type: credentialType, start_after: startAfter, limit },
    });
  }

  async getAdmin(): Promise<AdminResponse> {
    return this.client.queryContractSmart(this.contractAddress, { admin: {} });
  }
}

export class IssuerRegistryExecuteClient extends IssuerRegistryClient {
  private readonly signingClient: SigningCosmWasmClient;
  private readonly sender: string;

  constructor(
    signingClient: SigningCosmWasmClient,
    sender: string,
    contractAddress: string,
  ) {
    super(signingClient, contractAddress);
    this.signingClient = signingClient;
    this.sender = sender;
  }

  // ── Execute ────────────────────────────────────────────────────

  async registerIssuer(
    issuer: string,
    name: string,
    description: string,
    credentialTypes: string[],
    fee: "auto" | number = "auto",
  ): Promise<ExecuteResult> {
    return this.signingClient.execute(
      this.sender,
      this.contractAddress,
      {
        register_issuer: {
          issuer,
          name,
          description,
          credential_types: credentialTypes,
        },
      },
      fee,
    );
  }

  async revokeIssuer(
    issuer: string,
    reason: string,
    fee: "auto" | number = "auto",
  ): Promise<ExecuteResult> {
    return this.signingClient.execute(
      this.sender,
      this.contractAddress,
      { revoke_issuer: { issuer, reason } },
      fee,
    );
  }

  async updateIssuer(
    issuer: string,
    updates: {
      name?: string;
      description?: string;
      credential_types?: string[];
    },
    fee: "auto" | number = "auto",
  ): Promise<ExecuteResult> {
    return this.signingClient.execute(
      this.sender,
      this.contractAddress,
      {
        update_issuer: {
          issuer,
          ...updates,
        },
      },
      fee,
    );
  }

  async updateAdmin(
    newAdmin: string,
    fee: "auto" | number = "auto",
  ): Promise<ExecuteResult> {
    return this.signingClient.execute(
      this.sender,
      this.contractAddress,
      { update_admin: { new_admin: newAdmin } },
      fee,
    );
  }
}

File: packages/proofkit-types/src/issuer_registry.rs
use cosmwasm_schema::{cw_serde, QueryResponses};
use cosmwasm_std::Addr;

// ── Messages ──────────────────────────────────────────────────────────

#[cw_serde]
pub struct InstantiateMsg {
    /// Optional admin address; defaults to the instantiator.
    pub admin: Option<String>,
}

#[cw_serde]
pub enum ExecuteMsg {
    /// Register a new trusted credential issuer (admin only).
    RegisterIssuer {
        /// The issuer's on-chain address.
        issuer: String,
        /// Human-readable name for the issuer.
        name: String,
        /// Description of the issuer organization.
        description: String,
        /// Credential types this issuer is authorized to validate.
        credential_types: Vec<String>,
    },
    /// Revoke a previously registered issuer (admin only).
    RevokeIssuer {
        issuer: String,
        reason: String,
    },
    /// Update issuer details (admin only).
    UpdateIssuer {
        issuer: String,
        name: Option<String>,
        description: Option<String>,
        credential_types: Option<Vec<String>>,
    },
    /// Update the admin address (admin only).
    UpdateAdmin {
        new_admin: String,
    },
}

#[cw_serde]
#[derive(QueryResponses)]
pub enum QueryMsg {
    /// Check if an issuer is authorized for a specific credential type.
    #[returns(IsAuthorizedResponse)]
    IsAuthorized {
        issuer: String,
        credential_type: String,
    },
    /// Get details of a specific issuer.
    #[returns(IssuerResponse)]
    Issuer {
        issuer: String,
    },
    /// List all registered issuers.
    #[returns(IssuersResponse)]
    ListIssuers {
        start_after: Option<String>,
        limit: Option<u32>,
    },
    /// List issuers authorized for a specific credential type.
    #[returns(IssuersResponse)]
    IssuersByType {
        credential_type: String,
        start_after: Option<String>,
        limit: Option<u32>,
    },
    /// Get the current admin address.
    #[returns(AdminResponse)]
    Admin {},
}

// ── State types ───────────────────────────────────────────────────────

#[cw_serde]
pub struct IssuerRecord {
    pub address: Addr,
    pub name: String,
    pub description: String,
    pub credential_types: Vec<String>,
    pub registered_at: u64,
    pub active: bool,
    pub revoked_at: Option<u64>,
    pub revocation_reason: Option<String>,
}

// ── Response types ────────────────────────────────────────────────────

#[cw_serde]
pub struct IsAuthorizedResponse {
    pub authorized: bool,
    pub issuer_name: Option<String>,
}

#[cw_serde]
pub struct IssuerResponse {
    pub issuer: IssuerRecord,
}

#[cw_serde]
pub struct IssuersResponse {
    pub issuers: Vec<IssuerRecord>,
}

#[cw_serde]
pub struct AdminResponse {
    pub admin: Addr,
}
